mautic/mautic
# [WIP] do not depend on node_modules (and npm) during asset generation

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

in #12234 and follow-up PR's the hard coded (and outdated) js dependencies were switched to an package manager approach via NPM. This is really good, and this works during development, but unfortunately this brings some problems in other situations.

* npm is required for a composerized vanilla mautic installation, otherwise the assets can't be generated
* production envs where plugins are installed that add extra css / js, also require `mautic:assets:generate`, which required the `node_modules` folder, which requires npm being executed

This PR addresses this by adding an extra `postinstall` when running npm actions, to ensure the needed dependencies are copied to the `media` folder.

It's inspired on how Drupal does this, see https://git.drupalcode.org/project/drupal/-/blob/11.x/core/scripts/js/vendor-update.js and https://git.drupalcode.org/project/drupal/-/tree/11.x/core/assets/vendor



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. run `composer update --lock` or `ddev exec composer -- update --lock` to trigger the `post-update-cmd`
3. verify the ckeditor in Mautic is still working, e.g. by using the `send email` action on a contact
